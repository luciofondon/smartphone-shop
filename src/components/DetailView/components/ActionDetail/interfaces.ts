import { SelectChangeEvent } from '@mui/material';
import { IProductAdd } from '../../../Root/interfaces';
import { IProductDetail } from '../../interfaces';

export interface IActionDetail {
  product: IProductDetail;
  handleClickAddProduct: (product: IProductAdd) => void;
}

export interface IUseActionDetail {
  product: IProductDetail;
  handleClickAddProduct: (product: IProductAdd) => void;
}

export interface IUseActionDetailResponse {
  states: {
    storage: number;
    color: number;
  };
  actions: {
    handleSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
    handleChange: (event: SelectChangeEvent) => void;
  };
}
