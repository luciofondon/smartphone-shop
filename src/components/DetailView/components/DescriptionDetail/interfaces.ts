import { IProductDetail } from '../../interfaces';

export interface IDescriptionDetail {
  product: IProductDetail;
}
