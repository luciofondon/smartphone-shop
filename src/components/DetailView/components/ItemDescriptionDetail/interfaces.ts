export interface IItemDescriptrionDetail {
  label: string;
  value: string;
}
