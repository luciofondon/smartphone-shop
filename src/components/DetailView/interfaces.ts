import { IProductAdd } from '../Root/interfaces';

interface IColor {
  code: number;
  name: string;
}

interface IStorage {
  code: number;
  name: string;
}

export interface IProductDetail {
  announced: string;
  audioJack: string;
  battery: string;
  bluetooth: string[];
  brand: string;
  chipset: string;
  colors: string[];
  cpu: string;
  dimentions: string;
  displayResolution: string;
  displaySize: string;
  displayType: string;
  edge: string;
  externalMemory: string;
  gprs: string;
  gps: string;
  gpu: string;
  id: string;
  imgUrl: string;
  internalMemory: string[];
  model: string;
  networkSpeed: string;
  networkTechnology: string;
  nfc: string;
  options: { colors: IColor[]; storages: IStorage[] };
  os: string;
  price: string;
  primaryCamera: string[];
  radio: string;
  ram: string;
  secondaryCmera: string[];
  sensors: string[];
  sim: string;
  speaker: string;
  status: string;
  usb: string;
  weight: string;
  wlan: string[];
}

export interface IUseDetailViewResponse {
  states: {
    product: IProductDetail | null;
  };
  actions: {
    handleClickAddProduct: (product: IProductAdd) => void;
  };
}
