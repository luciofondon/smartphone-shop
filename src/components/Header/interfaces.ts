export interface IHeader {
  totalProducts: number;
}
