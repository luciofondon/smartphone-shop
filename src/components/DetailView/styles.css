.product-detail {
    padding-top: 80px;
    height: 600px;
    width: 60%;
    margin: auto;
}

.info-detail {
    float: right;
    width: 60%;
}

.img-detail {
    float: left;
    width: 20%;
}