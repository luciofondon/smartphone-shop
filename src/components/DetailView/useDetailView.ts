import { useContext, useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';
import { IProductAdd } from '../Root/interfaces';
import { ProductContext } from '../Root/Root';

import { IProductDetail, IUseDetailViewResponse } from './interfaces';

const useDetailView = (): IUseDetailViewResponse => {
  const [product, setProduct] = useState<IProductDetail | null>(null);
  const { handleAdd } = useContext(ProductContext);
  const { productId } = useParams();

  const handleClickAddProduct = (product: IProductAdd) => {
    handleAdd(product);
  };

  useEffect(() => {
    (async () => {
      const data = await fetch(`https://front-test-api.herokuapp.com/api/product/${productId}`);
      setProduct(await data.json());
    })();
  }, []);

  return {
    states: {
      product,
    },
    actions: {
      handleClickAddProduct,
    },
  };
};

export default useDetailView;
