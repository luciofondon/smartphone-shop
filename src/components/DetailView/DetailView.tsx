import React, { useContext } from 'react';

import ActionDetail from './components/ActionDetail';
import DescriptionDetail from './components/DescriptionDetail';
import useDetailView from './useDetailView';
import './styles.css';
import Loading from './components/Loading';

const DetailView = () => {
  const {
    states: { product },
    actions: { handleClickAddProduct },
  } = useDetailView();

  if (!product) {
    return <Loading />;
  }
  return (
    <div className="product-detail">
      {product && (
        <>
          <div className="img-detail">
            <img src={product.imgUrl} />
          </div>
          <div className="info-detail">
            <DescriptionDetail product={product} />
            <ActionDetail product={product} handleClickAddProduct={handleClickAddProduct} />
          </div>
        </>
      )}
    </div>
  );
};

export default DetailView;
