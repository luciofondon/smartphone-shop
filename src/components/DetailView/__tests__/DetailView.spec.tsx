import React from 'react';
import { render } from '@testing-library/react';

import DetailView from '../DetailView';
import { IProductDetail } from '../interfaces';
import useDetailView from '../useDetailView';

const productMock: IProductDetail = {
  announced: '',
  audioJack: '',
  battery: 'Removable Li-Po 4080 mAh battery',
  bluetooth: [],
  brand: 'Alcatel',
  chipset: '',
  colors: [],
  cpu: 'Octa-core 1.3 GHz Cortex-A53',
  dimentions: '153.8 x 75.6 x 8.5 mm (6.06 x 2.98 x 0.33 in)',
  displayResolution: '5.5 inches ',
  displaySize: '',
  displayType: '',
  edge: '',
  externalMemory: '',
  gprs: '',
  gps: '',
  gpu: '',
  id: '',
  imgUrl: 'https://front-test-api.herokuapp.com/images/P2oqviM96_ozwsgZkj9Xf.jpg',
  internalMemory: [],
  model: 'Liquid Z6 Plus',
  networkSpeed: '',
  networkTechnology: '',
  nfc: '',
  options: {
    colors: [{ code: 1, name: 'Rojo' }],
    storages: [{ code: 1, name: '100GB' }],
  },
  os: '',
  price: '220',
  primaryCamera: ['13 MPautofocusLED flash'],
  radio: '',
  ram: '3 GB RAM',
  secondaryCmera: ['5 MP'],
  sensors: [],
  sim: '',
  speaker: '',
  status: '',
  usb: '',
  weight: '169',
  wlan: [],
};

jest.mock('../useDetailView');

/* jest.mock('../useDetailView', () => () => ({
    states: {
        product: productMock
    },
    actions: { handleClickAddProduct: jest.fn() }
}
));
 */

test('render DetailView with Loading', () => {
  useDetailView.mockImplementation(() => ({
    states: {
      product: null,
    },
    actions: { handleClickAddProduct: jest.fn() },
  }));

  const { container, getByText } = render(<DetailView />);
  expect(getByText('Cargando...')).toBeTruthy();
  expect(container).toMatchSnapshot('empty');
});

test('render DetailView with Product', () => {
  useDetailView.mockImplementation(() => ({
    states: {
      product: productMock,
    },
    actions: { handleClickAddProduct: jest.fn() },
  }));
  const { container } = render(<DetailView />);
  expect(container).toMatchSnapshot('data');
});
