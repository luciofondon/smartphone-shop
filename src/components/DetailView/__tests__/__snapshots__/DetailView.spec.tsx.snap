// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render DetailView with Loading: empty 1`] = `
<div>
  <div
    class="loading"
  >
    Cargando...
  </div>
</div>
`;

exports[`render DetailView with Product: data 1`] = `
<div>
  <div
    class="product-detail"
  >
    <div
      class="img-detail"
    >
      <img
        src="https://front-test-api.herokuapp.com/images/P2oqviM96_ozwsgZkj9Xf.jpg"
      />
    </div>
    <div
      class="info-detail"
    >
      <div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Marca
          </div>
          <div
            class="value-description"
          >
            Alcatel
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Modelo
          </div>
          <div
            class="value-description"
          >
            Liquid Z6 Plus
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Precio
          </div>
          <div
            class="value-description"
          >
            220
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            CPU
          </div>
          <div
            class="value-description"
          >
            Alcatel
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            RAM
          </div>
          <div
            class="value-description"
          >
            Octa-core 1.3 GHz Cortex-A53
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Sistema operativo
          </div>
          <div
            class="value-description"
          />
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Resolucion
          </div>
          <div
            class="value-description"
          >
            5.5 inches 
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Batería
          </div>
          <div
            class="value-description"
          >
            Removable Li-Po 4080 mAh battery
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Cámara principal
          </div>
          <div
            class="value-description"
          >
            13 MPautofocusLED flash
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Cámara secundaria
          </div>
          <div
            class="value-description"
          >
            5 MP
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Cámara secundaria
          </div>
          <div
            class="value-description"
          >
            5 MP
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Dimensiones
          </div>
          <div
            class="value-description"
          >
            153.8 x 75.6 x 8.5 mm (6.06 x 2.98 x 0.33 in)
          </div>
        </div>
        <div
          class="item-description"
        >
          <div
            class="title-description"
          >
            Peso
          </div>
          <div
            class="value-description"
          >
            169
          </div>
        </div>
      </div>
      <div
        class="action-form"
      >
        <form>
          <div
            class="form-select"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                id="storage-label"
              >
                Almacenamiento
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="storage-label mui-component-select-storage"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-component-select-storage"
                  role="button"
                  tabindex="0"
                >
                  100GB
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  name="storage"
                  tabindex="-1"
                  value="1"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      Almacenamiento
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="form-select"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                id="color-label"
              >
                Color
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="color-label mui-component-select-color"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-component-select-color"
                  role="button"
                  tabindex="0"
                >
                  Rojo
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  name="color"
                  tabindex="-1"
                  value="1"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-14lo706"
                  >
                    <span>
                      Color
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            Añadir a la cesta
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </div>
  </div>
</div>
`;
